In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

## People.csv

In [4]:
data = pd.read_csv('people.csv')

In [5]:
data.head(5)

people_id  char_1      group_1  char_2        date   char_3   char_4  \
0     ppl_100  type 2  group 17304  type 2  2021-06-29   type 5   type 5   
1  ppl_100002  type 2   group 8688  type 3  2021-01-06  type 28   type 9   
2  ppl_100003  type 2  group 33592  type 3  2022-06-10   type 4   type 8   
3  ppl_100004  type 2  group 22593  type 3  2022-07-20  type 40  type 25   
4  ppl_100006  type 2   group 6534  type 3  2022-07-27  type 40  type 25   

   char_5  char_6   char_7   ...   char_29 char_30 char_31 char_32 char_33  \
0  type 5  type 3  type 11   ...     False    True    True   False   False   
1  type 5  type 3  type 11   ...     False    True    True    True    True   
2  type 5  type 2   type 5   ...     False   False    True    True    True   
3  type 9  type 4  type 16   ...      True    True    True    True    True   
4  type 9  type 3   type 8   ...     False   False    True   False   False   

  char_34 char_35 char_36 char_37 char_38  
0    True    True    True   False      36  
1    True    True    True   False      76  
2    True   False    True    True      99  
3    True    True    True    True      76  
4   False    True    True   False      84  

[5 rows x 41 columns]

In [6]:
data.shape

(189118, 41)

In [7]:
col_nunique = {}
for c in data.columns:
    col_nunique[c] = data[c].nunique()
    print('Col %s nunique: %d' % (c,data[c].nunique()))


Col people_id nunique: 189118
Col char_1 nunique: 2
Col group_1 nunique: 34224
Col char_2 nunique: 3
Col date nunique: 1196
Col char_3 nunique: 43
Col char_4 nunique: 25
Col char_5 nunique: 9
Col char_6 nunique: 7
Col char_7 nunique: 25
Col char_8 nunique: 8
Col char_9 nunique: 9
Col char_10 nunique: 2
Col char_11 nunique: 2
Col char_12 nunique: 2
Col char_13 nunique: 2
Col char_14 nunique: 2
Col char_15 nunique: 2
Col char_16 nunique: 2
Col char_17 nunique: 2
Col char_18 nunique: 2
Col char_19 nunique: 2
Col char_20 nunique: 2
Col char_21 nunique: 2
Col char_22 nunique: 2
Col char_23 nunique: 2
Col char_24 nunique: 2
Col char_25 nunique: 2
Col char_26 nunique: 2
Col char_27 nunique: 2
Col char_28 nunique: 2
Col char_29 nunique: 2
Col char_30 nunique: 2
Col char_31 nunique: 2
Col char_32 nunique: 2
Col char_33 nunique: 2
Col char_34 nunique: 2
Col char_35 nunique: 2
Col char_36 nunique: 2
Col char_37 nunique: 2
Col char_38 nunique: 101


In [8]:
for c in data.columns:
    if col_nunique[c]>10:
        continue
    
    print c
    print data[c].value_counts()

char_1
type 2    173867
type 1     15251
Name: char_1, dtype: int64
char_2
type 3    96553
type 2    77314
type 1    15251
Name: char_2, dtype: int64
char_5
type 9    63460
type 5    28896
type 4    28458
type 2    22105
type 8    11138
type 3    11032
type 7     9943
type 6     9189
type 1     4897
Name: char_5, dtype: int64
char_6
type 1    67051
type 3    50136
type 4    41966
type 5    17883
type 2     8537
type 6     1810
type 7     1735
Name: char_6, dtype: int64
char_8
type 3    76353
type 2    62773
type 1    12471
type 4    12264
type 6     8679
type 8     6002
type 5     5524
type 7     5052
Name: char_8, dtype: int64
char_9
type 3    80724
type 2    50729
type 4    13468
type 1     8985
type 8     8464
type 6     8123
type 9     7530
type 5     5845
type 7     5250
Name: char_9, dtype: int64
char_10
False    141660
True      47458
Name: char_10, dtype: int64
char_11
False    148363
True      40755
Name: char_11, dtype: int64
char_12
False    143664
True      45454
Name: char

In [80]:
data.describe()

char_38
count  189118.000000
mean       50.327399
std        35.752127
min         0.000000
25%        10.000000
50%        58.000000
75%        83.000000
max       100.000000

## Train

In [81]:
train = pd.read_csv('act_train.csv')

In [82]:
train.shape

(2197291, 15)

In [83]:
train.head()

people_id   activity_id        date activity_category char_1 char_2 char_3  \
0   ppl_100  act2_1734928  2023-08-26            type 4    NaN    NaN    NaN   
1   ppl_100  act2_2434093  2022-09-27            type 2    NaN    NaN    NaN   
2   ppl_100  act2_3404049  2022-09-27            type 2    NaN    NaN    NaN   
3   ppl_100  act2_3651215  2023-08-04            type 2    NaN    NaN    NaN   
4   ppl_100  act2_4109017  2023-08-26            type 2    NaN    NaN    NaN   

  char_4 char_5 char_6 char_7 char_8 char_9  char_10  outcome  
0    NaN    NaN    NaN    NaN    NaN    NaN  type 76        0  
1    NaN    NaN    NaN    NaN    NaN    NaN   type 1        0  
2    NaN    NaN    NaN    NaN    NaN    NaN   type 1        0  
3    NaN    NaN    NaN    NaN    NaN    NaN   type 1        0  
4    NaN    NaN    NaN    NaN    NaN    NaN   type 1        0

In [84]:
for c in train.columns:
    print('Column %s nunique: %d' % (c, train[c].nunique()))

Column people_id nunique: 151295
Column activity_id nunique: 2197291
Column date nunique: 411
Column activity_category nunique: 7
Column char_1 nunique: 51
Column char_2 nunique: 32
Column char_3 nunique: 11
Column char_4 nunique: 7
Column char_5 nunique: 7
Column char_6 nunique: 5
Column char_7 nunique: 8
Column char_8 nunique: 18
Column char_9 nunique: 19
Column char_10 nunique: 6515
Column outcome nunique: 2


In [85]:
train.outcome.value_counts()

0    1221794
1     975497
Name: outcome, dtype: int64

## Test

In [86]:
test = pd.read_csv('act_test.csv')

In [87]:
test.shape

(498687, 14)

In [88]:
test.head()

people_id   activity_id        date activity_category   char_1   char_2  \
0  ppl_100004   act1_249281  2022-07-20            type 1   type 5  type 10   
1  ppl_100004   act2_230855  2022-07-20            type 5      NaN      NaN   
2   ppl_10001   act1_240724  2022-10-14            type 1  type 12   type 1   
3   ppl_10001    act1_83552  2022-11-27            type 1  type 20  type 10   
4   ppl_10001  act2_1043301  2022-10-15            type 5      NaN      NaN   

   char_3  char_4  char_5  char_6  char_7   char_8   char_9    char_10  
0  type 5  type 1  type 6  type 1  type 1   type 7   type 4        NaN  
1     NaN     NaN     NaN     NaN     NaN      NaN      NaN   type 682  
2  type 5  type 4  type 6  type 1  type 1  type 13  type 10        NaN  
3  type 5  type 4  type 6  type 1  type 1   type 5   type 5        NaN  
4     NaN     NaN     NaN     NaN     NaN      NaN      NaN  type 3015

In [89]:
for c in test.columns:
    print('Column %s nunique: %d' % (c, test[c].nunique()))

Column people_id nunique: 37823
Column activity_id nunique: 498687


KeyboardInterrupt: 

## Check relations

In [ ]:
people_ids = set(data.people_id)
train_people_ids = set(train.people_id)
test_people_ids = set(test.people_id)

In [ ]:
len(people_ids)

In [ ]:
len(train_people_ids)

In [ ]:
len(test_people_ids)

In [ ]:
train_people_ids - people_ids

In [ ]:
test_people_ids - people_ids

In [ ]:
float(len(train_people_ids))/train.shape[0]

In [ ]:
1/_

In [90]:
train_people_ids & test_people_ids

set()

In [ ]:
print train.groupby('people_id').size().sort_values(ascending=False).head()
print train.groupby('people_id').size().median()
print train.groupby('people_id').size().mean()

In [ ]:
float(train.groupby('people_id').size().sort_values(ascending=False).head(100).sum())/train.shape[0]

In [ ]:
float(len(test_people_ids))/test.shape[0]

In [ ]:
1/_

In [ ]:
print test.groupby('people_id').size().sort_values(ascending=False).head()
print test.groupby('people_id').size().median()
print test.groupby('people_id').size().mean()

In [ ]:
float(test.groupby('people_id').size().sort_values(ascending=False).head(1000).sum())/test.shape[0]

In [ ]:
test.describe()

In [ ]:
train.describe()

In [ ]:
train.isnull().sum().astype(float)/train.shape[0]

In [ ]:
test.isnull().sum()

In [ ]:
test.isnull().sum().astype(float)/test.shape[0]

### Conclusions
- All samples in training and test dataset have entries in people
- Train dataset has median 5 people samples, average 14.5 and max 55k
- Top10 train samples represent 10% of people IDs, top 100 represent 11%. So top 10 maybe extreme values and should maybe be removed.
- Test dataset has median 5 people samples, average 13.2 and max 1640
- Top100 test samples represents 6% of samples, Top10 1%, Top1000 27%
- Test and train char_1 - char_9 contain approx 92% NaNs, char_10 only 8%
- There are no common people_ids between train and test